drivers: model kconfig for Ethernet drivers

leandrolanzieri opened the pull request:
### Contribution description

This models all Ethernet drivers in Kconfig. To test this (and also to make tests stack-independent) all related test applications are modified to run only using netdev (note that to test with GNRC we still have `tests/driver_netdev_common`). Additionally, specific test applications for `stm32_eth` and `esp_eth` are added.

~~**Note**: For now this includes some extra changes (e.g., 4b3031443a82bcb78216c4bc16335f026a8bbb21, fd9466b5f4c0f1f1357e3218820e995f1268dbd0, 8da03c2b04e7210cb2639ca615ecf00c4e93e1e0) that most likely will be split.~~

### Testing procedure
- Green CI

### Issues/PRs references
#17744
#17745
#17746
#17747

leandrolanzieri commented:
Now that the related PRs are merged I rebased to master. I will run the CI to see how it looks. Note that as I removed the `Makefile.ci` of the tests we will be probably be getting some overflows, but the list will hopefully be smaller, as we don't pull the stack anymore...

In the meantime, I'd appreciate some feedback on the test changes. Although we are gaining slimmer and simpler tests for the drivers, we would not be testing the auto inits. Those would be tested only on `tests/driver_netdev_common` for the devices that are default on some boards. @MrKevinWeiss proposed offline to create virtual boards to overcome this issue.

MrKevinWeiss commented:
It would be good to get some feedback on if the changes to the tests make sense.  This would mean that the autoinit code for the gnrc stack is not tested (though I think that should be tested somewhere else, maybe even with one of these virtual boards to not load down the CI). Probably this would mean we should have another PR that provides these tests.

If @benpicco or @kaspar030 can comment on the overall idea of the test changes that would be great.

leandrolanzieri commented on cpu/esp32/esp-eth/Kconfig:
```diff
@@ -0,0 +1,24 @@
+# directory for more details.
+#
+
+config MODULE_ESP_ETH
```
@gschorcht I could not find a dependency on a feature like `periph_eth` in the `Makefile.dep`, but I observe compilation errors on boards that do not provide a configuration. Should this module depend on `periph_eth` (which would be only provided when a config is present)?

gschorcht replied:
The board defines it as feature here: https://github.com/RIOT-OS/RIOT/blob/bec2868223b3559bdd969de5cd090a0c279baea5/boards/esp32-olimex-evb/Makefile.features#L15 Maybe this is wrong.

leandrolanzieri commented on cpu/esp32/esp-eth/Kconfig:
```diff
@@ -0,0 +1,24 @@
+# directory for more details.
+#
+
+config MODULE_ESP_ETH
```
@gschorcht I could not find a dependency on a feature like `periph_eth` in the `Makefile.dep`, but I observe compilation errors on boards that do not provide a configuration. Should this module depend on `periph_eth` (which would be only provided when a config is present)?

MrKevinWeiss commented:
> If @benpicco or @kaspar030 can comment on the overall idea of the test changes that would be great.

ping.

a quick yes or no to changing the driver tests to not include the network stack.

MrKevinWeiss commented on tests/external_board_dirs/netdev-ci-boards/nrf52840-ci-eth/Makefile.features:
```diff
@@ -0,0 +1,8 @@
+CPU_MODEL = nrf52840xxaa
```
Since the static tests fail and this is just a copy of an existing board maybe we do what was done on the esp:
```suggestion
include $(RIOTBOARD)/nrf52840dk/Makefile.features
```

maribu commented on tests/driver_enc28j60/commands.c:
```diff
@@ -0,0 +1,104 @@
+ * @{
+ *
+ * @file
+ * @brief       Shell commands for netdev_eth test application
```
Maybe this could be turned into a utility module? That would also allow using other applications to use this for debugging. I'm also personally not a fan of symlinking source files for code dedulication, as it makes it non-obvious that a seemingly application local file is relevant to other apps. (Also the `can_fast_ci_run.py` can be fooled by symlinks. But I guess this should be considered a bug of that tool.)

leandrolanzieri replied:
> I'm also personally not a fan of symlinking source files for code dedulication, as it makes it non-obvious that a seemingly application local file is relevant to other apps.

I'm actually following the current way. We have a `tests/driver_netdev_common` which is used to test almost all network devices, via symbolic links. 

So we would have a module for Ethernet, one for IEEE 802.15.4, and so on?

maribu replied:
OK. Let's keep it consistent. (And maybe a future PR can convert them all to modules.)

maribu commented on tests/driver_enc28j60/commands.c:
```diff
@@ -0,0 +1,104 @@
+static int cmd_print_addr(int argc, char **argv)
+{
+    (void) argc;
+    (void) argv;
```
```suggestion
    (void)argv;
```

A cast should IMO be styled like a unary operator. And we have no space after a unary operator, so that one can immediately see that `*ptr` is dereferencing `ptr` while `a * b` is multiplication, despite both operators use the same char.

maribu commented on tests/driver_enc28j60/commands.c:
```diff
@@ -0,0 +1,104 @@
+    int res;
+
+    if (3 != argc) {
+        goto usage;
```
```suggestion
        return _print_usage(argv[0]);
```

?

maribu commented on tests/driver_stm32_eth/Makefile:
```diff
@@ -0,0 +1,19 @@
+
+FEATURES_REQUIRED += periph_eth
+# TODO: complete with other STM32 CPUs that have ethernet
+FEATURES_REQUIRED_ANY += cpu_stm32f2|cpu_stm32f7
```
Note that `FEATURES_REQUIRED += periph_eth` should only select STM32 MCUs that do have ethernet. So no need to filter out MCU families.

leandrolanzieri replied:
But then how to limit to stm32? ESP also has `periph_eth`.

maribu replied:
Ah, OK!

How about `FEATURES_REQUIRED += cpu_stm32` then?

leandrolanzieri replied:
True, much direct. I'll change!

leandrolanzieri replied:
On a second thought, the `cpu_stm32` feature should be required by the `stm32_eth` driver itself, so I'll add it there.

maribu reviewed (commented):
Maybe you could split out the improvements of the test applications into a second PR? The KConfig migration part should be pretty straight forward to review.

MrKevinWeiss commented on tests/external_board_dirs/netdev-ci-boards/nrf52840-ci-eth/Makefile.features:
```diff
@@ -0,0 +1,8 @@
+include $(RIOTBOARD)/nrf52840dk/Makefile.features
+
+include $(RIOTBOARD)/common/nrf52xxxdk/Makefile.features
+
+# Various other features (if any)
+FEATURES_PROVIDED += periph_pwm
+FEATURES_PROVIDED += periph_usbdev
+FEATURES_PROVIDED += vdd_lc_filter_reg0
```
```suggestion
include $(RIOTBOARD)/nrf52840dk/Makefile.features
```

This is what I meant.

leandrolanzieri commented on boards/esp32-ethernet-kit-v1_0/Kconfig:
```diff
@@ -20,4 +20,6 @@ config BOARD_ESP32_ETHERNET_KIT_V1_0
     select HAS_PERIPH_PWM
     select HAS_PERIPH_SPI if !MODULE_ESP_JTAG
 
+    select HAVE_ESP_ETH
```
It is intentional. We have been keeping the `HAS_` (shared with Makefile) from the `HAVE_` (Kconfig only) features split. I'd like to keep it consistent if you don't mind.

maribu replied:
OK, makes sense :)

leandrolanzieri commented on tests/driver_dose/Makefile:
```diff
@@ -1,3 +1,12 @@
+USEMODULE += ps
+USEMODULE += shell
+USEMODULE += shell_commands
+
```
So, in order to be able to test the modelling a big part of the PR is adapting the tests to stop using the full network stack. Instead, we directly use the drivers via their netdev interface, for that I chose to use event threads. I'll split the test changes in their own PR then.

leandrolanzieri commented on boards/esp32-olimex-evb/Kconfig:
```diff
@@ -21,4 +21,6 @@ config BOARD_ESP32_OLIMEX_EVB
     select HAS_PERIPH_CAN
     select HAS_PERIPH_IR
 
+    select HAVE_ESP_ETH
```
See above

leandrolanzieri commented:
> > Maybe you could split out the improvements of the test applications into a second PR?
> 
> @maribu I split those changes in #17813, also moved the common part to its own module as suggested.

With #17813 merged we can continue with this one.

MrKevinWeiss commented on tests/driver_sam0_eth/Kconfig:
```diff
@@ -0,0 +1,12 @@
+    bool
+    default y
+    depends on HAS_CPU_SAMD5X
+    depends on HAS_PERIPH_ETH # TODO: complete with other SAM0 CPUs that have ethernet
```
@benpicco has a good overview of the sam0 series.  Are there any more boards/cpus that may have the ethernet peripheral?

benpicco replied:
The whole SAM E5x line has Ethernet (it's SAM D5x with **e**thernet like the SAM R21 is SAM D21 with **r**adio).
But we only have the `same54-xpro` board in RIOT with a CPU from the line.

leandrolanzieri commented:
Rebased to resolve conflicts

MrKevinWeiss commented:
Hmm seems to be some xtimer/ztimer issues...

leandrolanzieri commented:
> Hmm seems to be some xtimer/ztimer issues...

It was my fault, poor rebasing. Let's see now

leandrolanzieri commented:
Seems to work

MrKevinWeiss commented:
Any blockers? This only touches kconfig so it should be fine if it green.  @maribu ?

MrKevinWeiss reviewed (approved):
Looks good from the coding side.  Lets see what murdock says then please remove the limiting build commit.  ACK.

leandrolanzieri commented:
Fixed whitespace issue

MrKevinWeiss commented:
Finally eh?
